docs/amber.md

```diff
@@ -0,0 +1,138 @@
+For the CTS, the Vulkan engine is already configured and we just set the
+`VulkanEngineConfig` as needed.
+
+Accessing amber itself is done through the `Amber` class in `amber/amber.h`. We
+require the application to parse and execute a script as two separate steps.
+This separation allows for the scripts to be retrieved after a parse command
+and then pre-compiled and passed into the execution. Passing the compiled
+scripts is optional.
+
+### Delegate
+A delegate object can be provided to Amber, `nullptr` can be provided for the
+delegate. If the delegate is provided and the `LogGraphicsCalls` method returns
+`true` then the vulkan API wrappers will call `Log` for each call into Vulkan.
+If the `LogGraphicsCallsTime` also returns `true` then timings for those
+Vulkan calls will also be recorded. The timestamps are retrieved from the
+`GetTimestampNS` callback.
+
+### Buffer Extractions
+Amber can be instructed to retrieve the contents of buffers when execution is
+complete. This is done through the `extractions` list in the amber `Options`
+structure. You must set the `buffer_name` for each of the buffers you want to
+extract. When amber completes it will fill out the `width`, `height` and set
+of `Value` objects for that buffer.
+
+### Execution
+There are two methods to execute a parsed script `Execute` and
+`ExecuteWithShaderData`. They both accept the `Recipe` and `Options`, the
+`ExecuteWithShaderData` also accepts a map of shader name to data. The data
+is the compiled SPIR-V binary for the shader. This allows you to compile and
+cache the shader if needed.
+
+## Parsing component
+There are currently two parsing backends for Amber.
+[AmberScript](amber_script.md) and [VkScript](vk_script.md). The `AmberScript`
+parser will be used if the first 7 characters of the input buffer are
+`#!amber`, otherwise the `VkScript` parser will be used. The parsers both
+generate a `Script` which is our representation of the script file.
+
+### VkScript
+For VkScript we do a bit of work to make the script match AmberScript. A default
+pipeline is generated and all content in the script is considered part of the
+generated pipeline. We generate names for all of the buffers in the file. The
+framebuffer is named `framebuffer`. The generated depth buffer is named
+`depth_buffer`. For each buffer seen we generate a name of `AutoBuf-<num>` where
+the number is the current number of buffers seen counting from 0.
+
+The VkScript parser is broken into three major chunks. The `Parser`,
+`SectionParser` and `CommandParser`. The `Parser` is the overall driver for the
+parser. The `SectionParser` breaks the input file into the overall chunks (each
+of the sections separated by the \[blocks]). The `CommandParser` parses the
+`[test]` section specifically. For other sections they're parsed directly in
+the `Parser` object.
+
+### Parsed Representation
+The `Script` object is the owner for all of our data. Other objects hold
```

Comment:
"The `Script` object owns all script data."

(Surely it doesn't own _all_ the data.)

Reply:
Better?

The `Script` object owns all of the pipelines, buffers, shaders, and command
objects. Other objects hold pointers but the script holds the `unique_ptr` for
these objects.